File: parser/rule/signed/signed.go
/*
   The Fluent Programming Language
   -----------------------------------------------------
   This code is released under the GNU GPL v3 license.
   For more information, please visit:
   https://www.gnu.org/licenses/gpl-3.0.html
   -----------------------------------------------------
   Copyright (c) 2025 Rodrigo R. & All Fluent Contributors
   This program comes with ABSOLUTELY NO WARRANTY.
   For details type `fluent l`. This is free software,
   and you are welcome to redistribute it under certain
   conditions; type `fluent l -f` for details.
*/

package signed

import (
	"fluent/ast"
	"fluent/parser/error"
	"fluent/parser/queue"
	"fluent/parser/util"
	"fluent/token"
)

// Used to avoid building this string every time it is needed
var dummyOrString = "||"

// ProcessSignedOp processes a signed operation in the input tokens.
// It handles negation, operators, and expressions, and constructs an AST.
//
// Parameters:
// - input: A slice of tokens to be processed.
// - expressionRule: The rule representing an expression.
// - operatorRule: The rule representing an operator.
// - allowedSigns: A map of allowed token types for operators.
// - allowedSignsSlice: A slice of allowed token types for operators.
// - candidate: A pointer to an AST node representing the candidate expression.
// - exprQueue: A pointer to a slice of queue elements for expressions.
// - hasSkippedTokens: A boolean indicating if tokens have been skipped.
// - isBoolean: A boolean indicating if the expression is a boolean.
// - includeSigns: A boolean indicating if the signs should be included.
//
// Returns:
// - An AST representing the processed expression.
// - An error if the processing fails.
func ProcessSignedOp(
	input []token.Token,
	expressionRule ast.Rule,
	operatorRule ast.Rule,
	allowedSigns map[token.Type]struct{},
	allowedSignsSlice []token.Type,
	candidate *ast.AST,
	exprQueue *[]queue.Element,
	hasSkippedTokens bool,
	isBoolean bool,
	includeSigns bool,
) (*ast.AST, *error.Error) {
	// Check the input length
	if len(input) < 1 {
		return nil, &error.Error{
			Line:     candidate.Line,
			Column:   candidate.Column,
			File:     candidate.File,
			Expected: []ast.Rule{expressionRule},
		}
	}

	result := ast.AST{
		Rule:     expressionRule,
		Line:     input[0].Line,
		Column:   input[0].Column,
		File:     &input[0].File,
		Children: &[]*ast.AST{},
	}

	if candidate != nil && candidate.Rule != ast.Program {
		*result.Children = append(*result.Children, &ast.AST{
			Rule:     ast.Expression,
			Line:     candidate.Line,
			Column:   candidate.Column,
			File:     candidate.File,
			Children: &[]*ast.AST{candidate},
		})
	}

	// The expression parser has processed a part of the expression
	// Therefore, we receive all tokens from the operator and so on
	expectingOperator := hasSkippedTokens
	inNegation := false

	// Used to skip tokens
	skip := 0

	for i, unit := range input {
		if i < skip {
			continue
		}

		if unit.TokenType == token.Not {
			if !isBoolean {
				return nil, &error.Error{
					Line:     unit.Line,
					Column:   unit.Column,
					File:     &unit.File,
					Expected: []ast.Rule{expressionRule},
				}
			}

			if expectingOperator && hasSkippedTokens {
				return nil, &error.Error{
					Line:     unit.Line,
					Column:   unit.Column,
					File:     &unit.File,
					Expected: []ast.Rule{expressionRule},
				}
			}

			hasSkippedTokens = true
			inNegation = true
			expectingOperator = false

			// Add the operator to the result
			signValue := "!"
			*result.Children = append(*result.Children, &ast.AST{
				Rule:     operatorRule,
				Line:     unit.Line,
				Column:   unit.Column,
				File:     &unit.File,
				Value:    &signValue,
				Children: &[]*ast.AST{},
			})

			continue
		}

		if expectingOperator {
			if _, ok := allowedSigns[unit.TokenType]; !ok {
				return nil, &error.Error{
					Line:     unit.Line,
					Column:   unit.Column,
					File:     &unit.File,
					Expected: []ast.Rule{operatorRule},
				}
			}

			if includeSigns {
				signValue := ""

				// Determine the value to use for the operator
				switch unit.TokenType {
				case token.Plus:
					signValue = "+"
				case token.Minus:
					signValue = "-"
				case token.Asterisk:
					signValue = "*"
				case token.Slash:
					signValue = "/"
				case token.And:
					signValue = "&&"
				case token.Equal:
					signValue = "=="
				case token.NotEqual:
					signValue = "!="
				case token.GreaterThan:
					signValue = ">"
				case token.LessThan:
					signValue = "<"
				case token.GreaterThanOrEqual:
					signValue = ">="
				case token.LessThanOrEqual:
					signValue = "<="
				default:
				}

				// Add the operator to the result
				*result.Children = append(*result.Children, &ast.AST{
					Rule:     operatorRule,
					Line:     unit.Line,
					Column:   unit.Column,
					File:     &unit.File,
					Value:    &signValue,
					Children: &[]*ast.AST{},
				})
			}

			// Reset the flag
			expectingOperator = false
			continue
		}

		// Extract the tokens before the next operator
		extracted := util.ExtractTokensBefore(
			input[i:],
			allowedSignsSlice,
			true,
			token.OpenParen,
			token.CloseParen,
			false,
		)

		if extracted == nil {
			return nil, &error.Error{
				Line:     unit.Line,
				Column:   unit.Column,
				File:     &unit.File,
				Expected: []ast.Rule{expressionRule},
			}
		}

		firstElement := extracted[0]
		// Increment the skip
		skip = i + len(extracted)

		// Add the tokens to the expression queue
		expressionNode := ast.AST{
			Rule:     ast.Expression,
			Line:     firstElement.Line,
			Column:   firstElement.Column,
			File:     &firstElement.File,
			Children: &[]*ast.AST{},
		}

		if isBoolean {
			// Skip the tokens by OR for precedence
			split := util.SplitTokens(
				extracted,
				token.Or,
				[]token.Type{token.OpenParen},
				[]token.Type{token.CloseParen},
			)
			splitLen := len(split)

			// Used to handle nested boolean expressions
			var pushTo *ast.AST

			if splitLen > 1 {
				pushTo = &ast.AST{
					Rule:     ast.BooleanExpression,
					Line:     firstElement.Line,
					Column:   firstElement.Column,
					File:     &firstElement.File,
					Children: &[]*ast.AST{},
				}

				*expressionNode.Children = append(*expressionNode.Children, pushTo)
			}

			for j, tokens := range split {
				var trace *token.Token

				if len(tokens) == 0 {
					trace = &firstElement
				} else {
					trace = &tokens[0]
				}

				// Append the nested expression to the result
				if pushTo != nil {
					// Create a nested expression
					nestedExpression := ast.AST{
						Rule:     ast.Expression,
						Line:     trace.Line,
						Column:   trace.Column,
						File:     &firstElement.File,
						Children: &[]*ast.AST{},
					}

					*pushTo.Children = append(*pushTo.Children, &nestedExpression)

					*exprQueue = append(*exprQueue, queue.Element{
						Tokens: tokens,
						Parent: &nestedExpression,
					})
				} else {
					*exprQueue = append(*exprQueue, queue.Element{
						Tokens: tokens,
						Parent: &expressionNode,
					})
				}

				// Add the OR operator if it is not the last expression
				if j < splitLen-1 {
					orRule := ast.AST{
						Rule:     operatorRule,
						Line:     firstElement.Line,
						Column:   firstElement.Column,
						File:     &firstElement.File,
						Value:    &dummyOrString,
						Children: &[]*ast.AST{},
					}

					if pushTo != nil {
						*pushTo.Children = append(*pushTo.Children, &orRule)
					} else {
						*expressionNode.Children = append(*expressionNode.Children, &orRule)
					}
				}
			}
		} else {
			*exprQueue = append(*exprQueue, queue.Element{
				Tokens:       extracted,
				Parent:       &expressionNode,
				IsPropAccess: !includeSigns,
			})
		}

		// Append the expression to the result
		*result.Children = append(*result.Children, &expressionNode)

		// Set the flags
		inNegation = false
		expectingOperator = true
	}

	// The loop should never end not expecting an operator
	if !expectingOperator || inNegation {
		return nil, &error.Error{
			Line:     input[len(input)-1].Line,
			Column:   input[len(input)-1].Column,
			File:     &input[len(input)-1].File,
			Expected: []ast.Rule{expressionRule},
		}
	}

	// Used to skip tokens
	return &result, nil
}

File: parser/rule/signed/signed.cpp
/*
        ==== The Zelix Programming Language ====
---------------------------------------------------------
  - This file is part of the Zelix Programming Language
    codebase. Zelix is a fast, statically-typed and
    memory-safe programming language that aims to
    match native speeds while staying highly performant.
---------------------------------------------------------
  - Zelix is categorized as free software; you can
    redistribute it and/or modify it under the terms of
    the GNU General Public License as published by the
    Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.
---------------------------------------------------------
  - Zelix is distributed in the hope that it will
    be useful, but WITHOUT ANY WARRANTY; without even
    the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE. See the GNU General Public
    License for more details.
---------------------------------------------------------
  - You should have received a copy of the GNU General
    Public License along with Zelix. If not, see
    <https://www.gnu.org/licenses/>.
*/

//
// Created by rodri on 8/16/25.
//

#include "signed.h"
#include "parser/parser.h"
using namespace zelix;

template<bool Arithmetic>
inline parser::ast::rule_t rule(lexer::token *&token)
{
    if constexpr (Arithmetic)
    {
        switch (token->type)
        {
            case lexer::token::PLUS:
                return parser::ast::SUM;
            case lexer::token::MINUS:
                return parser::ast::SUB;
            case lexer::token::MULTIPLY:
                return parser::ast::MUL;
            case lexer::token::DIVIDE:
                return parser::ast::DIV;
            default:
                return parser::ast::ROOT; // Default case if no arithmetic operation matches
        }
    }

    switch (token->type)
    {
        case lexer::token::BOOL_EQ:
            return parser::ast::EQ;
        case lexer::token::BOOL_GT:
            return parser::ast::GT;
        case lexer::token::BOOL_GTE:
            return parser::ast::GTE;
        case lexer::token::BOOL_LT:
            return parser::ast::LT;
        case lexer::token::BOOL_LTE:
            return parser::ast::LTE;
        case lexer::token::BOOL_NEQ:
            return parser::ast::NEQ;
        default:
            return parser::ast::ROOT; // Default case if no arithmetic operation matches
    }
}

template<bool Arithmetic>
inline void process_op(
    lexer::token *&next,
    memory::lazy_allocator<parser::ast> &allocator,
    parser::ast *arithmetic_node
)
{
    // Create a new AST node for the arithmetic operation
    parser::ast *arithmetic_op_node = allocator.alloc();
    arithmetic_op_node->rule = rule<Arithmetic>(next);
    arithmetic_node->children.push_back(arithmetic_op_node); // Add the operation node
}

template<bool Arithmetic>
inline void process_sub(
    container::vector<lexer::token *> &current_tokens,
    lexer::token *&next,
    parser::ast *arithmetic_node,
    parser::ast *candidate,
    memory::lazy_allocator<parser::ast> &allocator,
    memory::lazy_allocator<parser::rule::expr::queue_node> &q_allocator,
    container::vector<parser::rule::expr::queue_node *>
    &expr_queue, bool &first_iteration,
    const bool append_sign = true
)
{
    // Handle first iteration
    if (first_iteration)
    {
        first_iteration = false; // Set the flag to false for subsequent iterations

        if (append_sign)
        {
            process_op<Arithmetic>(next, allocator,
                                   arithmetic_node); // Process the arithmetic operator
        }

        return;
    }

    // Append the current tokens to the last nested node
    if (current_tokens.empty())
    {
        parser::global_err.type = parser::UNEXPECTED_TOKEN;
        parser::global_err.column = next->column;
        parser::global_err.line = next->line;
        throw except::exception("Unexpected empty arithmetic expression");
    }

    // Create a new expression node for the current tokens
    parser::ast *expr_node = allocator.alloc();
    expr_node->rule = parser::ast::EXPRESSION;
    arithmetic_node->children.push_back(expr_node); // Add the expression node

    // Queue the current tokens for processing
    const auto q_el = q_allocator.alloc();
    q_el->tokens = container::stream<lexer::token *>(current_tokens);
    q_el->node = expr_node;
    expr_queue.push_back(q_el);
    current_tokens.clear(); // Clear the current tokens for the next operation

    if (!append_sign)
        return; // If we don't want to append the sign, return early
    process_op<Arithmetic>(next, allocator,
                           arithmetic_node); // Process the arithmetic operator
}

template<bool Arithmetic>
inline void process_high_precedence(
    container::vector<lexer::token *> &current_tokens,
    lexer::token *&next,
    parser::ast *&last_nested,
    parser::ast *&arithmetic_node,
    parser::ast *&candidate,
    memory::lazy_allocator<parser::ast> &allocator,
    memory::lazy_allocator<parser::rule::expr::queue_node> &q_allocator,
    container::vector<parser::rule::expr::queue_node *> &expr_queue,
    bool &first_iteration
)
{
    // Allocate the last nested node if it doesn't exist
    if (last_nested == nullptr)
    {
        last_nested = allocator.alloc();
        last_nested->rule = parser::ast::ARITHMETIC;
        arithmetic_node->children.push_back(last_nested);
    }

    process_sub<Arithmetic>(
        current_tokens,
        next,
        last_nested,
        candidate,
        allocator,
        q_allocator,
        expr_queue,
        first_iteration
    ); // Process the subexpression
}

template<bool Arithmetic>
inline void process_low_precedence(
    container::vector<lexer::token *> &current_tokens,
    lexer::token *&next,
    parser::ast *&last_nested,
    parser::ast *&arithmetic_node,
    parser::ast *&candidate,
    memory::lazy_allocator<parser::ast> &allocator,
    memory::lazy_allocator<parser::rule::expr::queue_node> &q_allocator,
    container::vector<parser::rule::expr::queue_node *> &expr_queue,
    bool &first_iteration
)
{
    // Check if we have a last nested node
    if (last_nested != nullptr)
    {
        // Process the last nested node
        process_sub<Arithmetic>(current_tokens, next, last_nested, candidate, allocator, q_allocator, expr_queue,
                                first_iteration,
                                false); // Process the subexpression
    }
    else
    {
        process_sub<Arithmetic>(current_tokens, next, arithmetic_node, candidate, allocator, q_allocator, expr_queue,
                                first_iteration,
                                false); // Process the subexpression
    }

    // Set the last nested node to nullptr
    last_nested = nullptr; // Reset the last nested node

    process_op<Arithmetic>(next, allocator,
                           arithmetic_node); // Process the arithmetic operator
}

template<bool Arithmetic>
parser::ast *parser::rule::signed_op(
    ast *&candidate,
    container::stream<lexer::token *> &tokens,
    memory::lazy_allocator<ast> &allocator,
    memory::lazy_allocator<expr::queue_node> &q_allocator,
    container::vector<expr::queue_node *> &expr_queue
)
{
    // Create the arithmetic AST node
    ast *arithmetic_node = allocator.alloc();
    arithmetic_node->children.push_back(candidate); // Add the candidate as the first child
    if constexpr (Arithmetic)
    {
        arithmetic_node->rule = ast::ARITHMETIC;
    }
    else
    {
        arithmetic_node->rule = ast::BOOLEAN;
    }

    ast *last_nested = nullptr; // Last nested arithmetic node
    auto next_opt = tokens.next();
    auto &next = next_opt.get();
    // Tokens for the current subexpression
    container::vector<lexer::token *> current_tokens;
    // Track nested parentheses
    size_t nested_count = 0;
    bool first_iteration = true; // Flag for the first iteration
    bool last_is_mul = false; // Track if the last operation was multiplication/division

    while (next_opt.is_some())
    {
        next = next_opt.get();

        if constexpr (Arithmetic)
        {
            // Check if we have to break
            if (next->type >= lexer::token::BOOL_EQ && next->type <= lexer::token::BOOL_GTE)
            {
                break;
            }
        }

        // Check if we have an arithmetic operation with high precedence
        bool high_precedence;
        if constexpr (Arithmetic)
        {
            high_precedence = next->type == lexer::token::MULTIPLY || next->type == lexer::token::DIVIDE;
        }
        else
        {
            high_precedence = next->type == lexer::token::OR;
        }

        bool low_precedence;
        if constexpr (Arithmetic)
        {
            low_precedence = next->type == lexer::token::PLUS || next->type == lexer::token::MINUS;
        }
        else
        {
            low_precedence = (next->type >= lexer::token::BOOL_EQ && next->type <= lexer::token::BOOL_GTE) ||
                             next->type == lexer::token::AND;
        }

        if (nested_count == 0 && high_precedence)
        {
            last_is_mul = true;
            process_high_precedence<Arithmetic>(
                current_tokens,
                next,
                last_nested,
                arithmetic_node,
                candidate,
                allocator,
                q_allocator,
                expr_queue,
                first_iteration
            ); // Process the multiplication/division operation
        }
        // Check if we have an arithmetic operator with low precedence
        else if (nested_count == 0 && low_precedence)
        {
            last_is_mul = false;
            process_low_precedence<Arithmetic>(
                current_tokens,
                next,
                last_nested,
                arithmetic_node,
                candidate,
                allocator,
                q_allocator,
                expr_queue,
                first_iteration
            ); // Process the addition/subtraction operation
        }
        // Push the token to the current vector
        else
        {
            // Check if we have a nested expression
            if (next->type == lexer::token::OPEN_PAREN)
            {
                nested_count++;

                // Make sure we don't include the opening parenthesis in the current tokens
                if (nested_count == 1)
                {
                    next_opt = tokens.next();
                    continue;
                }
            }
            else if (next->type == lexer::token::CLOSE_PAREN)
            {
                // Make sure we don't overflow
                if (nested_count == 0)
                {
                    global_err.type = UNEXPECTED_TOKEN;
                    global_err.column = next->column;
                    global_err.line = next->line;
                    throw except::exception("Unexpected closing parenthesis in arithmetic expression");
                }

                nested_count--;
                // Make sure we don't include the closing parenthesis in the current tokens
                if (nested_count == 0)
                {
                    next_opt = tokens.next();
                    continue;
                }
            }

            current_tokens.push_back(next); // Add the token to the current tokens
        }

        next_opt = tokens.next();
    }

    // In a valid expression, the last chunk should not be empty
    if (current_tokens.empty())
    {
        global_err.type = UNEXPECTED_TOKEN;
        global_err.column = next->column;
        global_err.line = next->line;
        throw except::exception("Unexpected empty arithmetic expression");
    }

    // Process the last arithmetic operation
    const auto last_sub = allocator.alloc();
    last_sub->rule = ast::EXPRESSION;
    if (last_is_mul)
    {
        last_nested->children.push_back(last_sub); // Add the last subexpression to the last nested node
    }
    else
    {
        arithmetic_node->children.push_back(last_sub); // Add the last subexpression to the arithmetic node
    }

    // Queue the current tokens for processing
    const auto q_el = q_allocator.alloc();
    q_el->tokens = container::stream<lexer::token *>(current_tokens); // Clone
    q_el->node = last_sub;
    expr_queue.push_back(q_el);

    // Unwrap if we have a nested expression
    if constexpr (Arithmetic)
    {
        if (arithmetic_node->children.size() == 1)
        {
            const auto first_child = arithmetic_node->children[0];
            allocator.dealloc(arithmetic_node); // Deallocate the arithmetic node
            arithmetic_node = first_child; // Set the first child as the new node
        }
    }

    return arithmetic_node;
}

// signed_op templates
template
parser::ast*
    parser::rule::signed_op<true>(
        ast*&,
        container::stream<lexer::token*>&,
        memory::lazy_allocator<ast, 256ul, false>&,
        memory::lazy_allocator<expr::queue_node, 256ul, false>&,
        container::vector<expr::queue_node*>&
    );

template
parser::ast*
    parser::rule::signed_op<false>(
        ast*&,
        container::stream<lexer::token*>&,
        memory::lazy_allocator<ast, 256ul, false>&,
        memory::lazy_allocator<expr::queue_node, 256ul, false>&,
        container::vector<expr::queue_node*>&
    );